### PR TITLE
Enhance NuGet Package Installation for C++ and C# Projects

### PR DESCRIPTION
**Pull request by ShashankNay**
This PR introduces improvements to the NuGet package installation process in our Visual Studio templates for both C++ and C# projects. Previously, package names were statically defined in `NuGetPackageList.cs`. This approach has been replaced by a more dynamic method, where package names are now passed via the `customparameters` in the `.vstemplate` files, allowing for greater flexibility and easier maintenance.

**Key enhancements include:**

- **Dynamic Package References**: Shift from the static `NuGetPackageList.cs` to `.vstemplate`-based `customparameters` for defining NuGet packages.
- **C# Multi-Package Installation**: Extend the capability of C# project templates to install multiple NuGet packages, leveraging the new dynamic parameters method.
- **Additional Package Support**: Added the "Microsoft.Windows.SDK.BuildTools" package to the list of NuGet packages for installation, expanding the development toolkit.
- **wapproj Timing Logic**: Addressed a timing issue in `wapproj` files, identified in the previous [PR ](https://github.com/microsoft/WindowsAppSDK/pull/4322#issuecomment-2048284112)`wapproj` has been found to align more closely with `csproj` behavior, prompting the adoption of C# installer logic – specifically, the use of `IVsNuGetProjectUpdateEvents` with `OnSolutionRestoreFinished`.

Looking ahead, we recognize the convergence in the installation logic between C# and C++ project templates. With this in mind, plans are underway to merge them into a unified, streamlined process.

### CODE REVIEW

**Comment by ghost1372** [comment]
Hi @ShashankNay 
Does this pr support offline installation?
Sometimes we dont have any access to internet for installing latest packages so we should use our local cached packages...
Currently if we want to create a new project  in a offline system, we get an error
Which wasdk can not be installed and we should install it manually.
I added a pr for fixing this before
https://github.com/microsoft/WindowsAppSDK/pull/4327

File: dev/VSIX/ProjectTemplates/Desktop/CSharp/PackagedApp/BlankApp/WinUI.Desktop.Cs.BlankApp.vstemplate [hunk 1]
```diff
@@ -34,15 +34,13 @@
       <ProjectItem ReplaceParameters="true" TargetFileName="App.xaml.cs">App.xaml.cs</ProjectItem>
       <ProjectItem ReplaceParameters="true" TargetFileName="MainWindow.xaml">MainWindow.xaml</ProjectItem>
       <ProjectItem ReplaceParameters="true" TargetFileName="MainWindow.xaml.cs">MainWindow.xaml.cs</ProjectItem>
-    </Project>
+    </Project>      
+    <CustomParameters>
+      <CustomParameter Name="$NuGetPackages$" Value="Microsoft.WindowsAppSDK;Microsoft.Windows.SDK.BuildTools"/>
```
Comment (Scottj1s) [inline dev/VSIX/ProjectTemplates/Desktop/CSharp/PackagedApp/BlankApp/WinUI.Desktop.Cs.BlankApp.vstemplate:39]: IOU: could just name id NuGetPackages (don't need the dollar signs)

**Review by Scottj1s — APPROVED** [review]
assuming we've capture the consolidation of identical wizard project and source files as an IOU, let's complete this

**Pull request merged**